aerogear/unifiedpush-operator
# Optionally create backup CronJob(s) based on UPS CR

JIRA: https://issues.jboss.org/browse/AEROGEAR-9433

----

# Verification steps

* Check out this PR
* Deploy the operator
```
make cluster/prepare
cat deploy/operator.yaml | sed -e "s|quay.io/aerogear/unifiedpush-operator:master|quay.io/aerogear/unifiedpush-operator:$(git rev-parse --short HEAD)|g" | kubectl create -n unifiedpush -f -
```
* Create a ServiceAccount called `backupjob`. This is needed for the resulting cronjobs, but is not created by the operator, since it may need additional access that the operator may not be able to give -- if the secrets being used are in different namespaces, then there would need to be relevant rolebindings for it, so we deliberately leave out the creation of all of that stuff:
```
kubectl create sa backupjob
```
* Create a UPS CR that has backups specified:
```
kubectl apply -f deploy/crds/push_v1alpha1_unifiedpushserver_cr_with_backup.yaml
```

* Verify that there are no persistent errors in the operator logs, and that everything gets created as expected. The important things to check for here are "cronjob" objects:

```
kubectl get cronjobs -l app=example-ups-with-backups
NAME                    SCHEDULE    SUSPEND   ACTIVE    LAST SCHEDULE   AGE
ups-daily-at-midnight   0 0 * * *   False     0         <none>          <invalid>
```

* Edit the schedule of the cronjob, and notice that it gets changed back immediately (note the output may not look exactly like this, due to concurrency, I've annotated with comments):
```
$ oc get cronjobs -l app=example-ups-with-backups -w & oc patch cronjob ups-daily-at-midnight -p '{"spec": {"schedule": "1 0 * * *"}}'
[1] 17231 #pid of backgrounded `oc get -w` task, you can kill it after this step, e.g. `kill 17231`
NAME                    SCHEDULE    SUSPEND   ACTIVE    LAST SCHEDULE   AGE
ups-daily-at-midnight   0 0 * * *   False     0         <none>          4m #before patch, as specified in CR
cronjob.batch/ups-daily-at-midnight patched
ups-daily-at-midnight   1 0 * * *   False     0         <none>    4m #after patch
ups-daily-at-midnight   0 0 * * *   False     0         <none>    4m #after reconcile fixes it back to what the CR specified
```

* Update the schedule for a backup in the CR, and notice that the cronjob gets updated immediately, as expected (just use `kubectl edit ups example-ups-with-backups` for this...I didn't spend long enough to figure out the `kubectl patch` for it).

@odra has already verified that the backups work as expected. If you want to do that also, you'll need to configure secrets (the one for AWS details at least, and the one for PGP if you want to verify that). It's not worth the effort, in my opinion, given that Leo has already done it (thanks Leo!).
